File: go/clients/tracked_openai.go
package clients

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/DexwoxBusiness/dexcost-go/core"
	"github.com/DexwoxBusiness/dexcost-go/pricing"
)

// TrackedOpenAI wraps an OpenAI-compatible client to automatically record
// LLM cost events for every chat completion call. The inner client is
// accepted as interface{} so that the dexcost SDK does not depend on any
// specific OpenAI Go SDK.
//
// The wrapper expects the inner client to expose a method with the signature:
//
//	CreateChatCompletion(ctx context.Context, req interface{}) (interface{}, error)
//
// If the inner client does not satisfy this interface, calls will return an error.
//
// Usage with sashabaranov/go-openai:
//
//	client := openai.NewClient(apiKey)
//	tracked := clients.NewTrackedOpenAI(client, tracker, pricingEngine)
//	resp, err := tracked.CreateChatCompletion(ctx, req)
type TrackedOpenAI struct {
	inner   interface{}
	tracker *core.Tracker
	pricing *pricing.Engine
}

// openaiCompletionCreator is the interface that the inner client must satisfy.
type openaiCompletionCreator interface {
	CreateChatCompletion(ctx context.Context, req interface{}) (interface{}, error)
}

// NewTrackedOpenAI creates a new TrackedOpenAI wrapper.
func NewTrackedOpenAI(inner interface{}, tracker *core.Tracker, pricing *pricing.Engine) *TrackedOpenAI {
	return &TrackedOpenAI{inner: inner, tracker: tracker, pricing: pricing}
}

// CreateChatCompletion calls the inner client's CreateChatCompletion method,
// records an llm_call event with cost/token data, and returns the response.
// If no task is present in ctx, an auto-task is created and finalized.
func (t *TrackedOpenAI) CreateChatCompletion(ctx context.Context, req interface{}) (interface{}, error) {
	start := time.Now()

	// Call the inner client.
	resp, err := t.callInner(ctx, req)

	latencyMs := int(time.Since(start).Milliseconds())

	// Determine task context.
	autoCreated := false
	var task *core.Task
	if existing := core.GetCurrentTask(ctx); existing != nil {
		task = existing
	} else {
		auto := core.CreateAutoTask(ctx, "openai.chat")
		task = &auto
		autoCreated = true
		if insErr := t.tracker.Buffer().InsertTask(*task); insErr != nil {
			log.Printf("[dexcost] failed to persist task: %v", insErr)
		}
	}

	if err != nil {
		// Record a failed event with zero cost.
		event := core.NewEvent(task.TaskID, core.EventTypeLLMCall)
		event.Provider = "openai"
		event.CostConfidence = core.CostConfidenceUnknown
		event.PricingSource = core.PricingSourceUnknown
		event.LatencyMs = &latencyMs
		event.Details["error"] = err.Error()
		if insErr := t.tracker.Buffer().InsertEvent(event); insErr != nil {
			log.Printf("[dexcost] failed to persist event: %v", insErr)
		}

		logEvent(&event, task.TaskType)

		if autoCreated {
			core.FinalizeAutoTask(task, &event, string(core.TaskStatusFailed), t.tracker.Buffer())
		}
		return nil, err
	}

	// Extract response fields via type assertion to map.
	respMap, mapErr := toResponseMap(resp)
	if mapErr != nil {
		// Response is not a map; record a minimal event.
		event := core.NewEvent(task.TaskID, core.EventTypeLLMCall)
		event.Provider = "openai"
		event.CostConfidence = core.CostConfidenceUnknown
		event.PricingSource = core.PricingSourceUnknown
		event.LatencyMs = &latencyMs
		if insErr := t.tracker.Buffer().InsertEvent(event); insErr != nil {
			log.Printf("[dexcost] failed to persist event: %v", insErr)
		}

		logEvent(&event, task.TaskType)

		if autoCreated {
			core.FinalizeAutoTask(task, &event, string(core.TaskStatusSuccess), t.tracker.Buffer())
		}
		return resp, nil
	}

	event, recordErr := RecordOpenAIResponse(t.tracker.Buffer(), t.pricing, task.TaskID, respMap)
	if recordErr != nil {
		// Fallback: record a minimal event.
		event = core.NewEvent(task.TaskID, core.EventTypeLLMCall)
		event.Provider = "openai"
		event.CostConfidence = core.CostConfidenceUnknown
		event.PricingSource = core.PricingSourceUnknown
		event.LatencyMs = &latencyMs
		if insErr := t.tracker.Buffer().InsertEvent(event); insErr != nil {
			log.Printf("[dexcost] failed to persist event: %v", insErr)
		}
	} else {
		event.LatencyMs = &latencyMs
		if updErr := t.tracker.Buffer().UpdateEvent(event); updErr != nil {
			log.Printf("[dexcost] failed to update event: %v", updErr)
		}
	}

	logEvent(&event, task.TaskType)

	if autoCreated {
		core.FinalizeAutoTask(task, &event, string(core.TaskStatusSuccess), t.tracker.Buffer())
	}

	return resp, nil
}

// callInner invokes the inner client's CreateChatCompletion method.
func (t *TrackedOpenAI) callInner(ctx context.Context, req interface{}) (interface{}, error) {
	if cc, ok := t.inner.(openaiCompletionCreator); ok {
		return cc.CreateChatCompletion(ctx, req)
	}
	return nil, fmt.Errorf("clients: inner OpenAI client does not implement CreateChatCompletion(context.Context, interface{}) (interface{}, error)")
}

// logEvent is a package-level helper that calls the dev console LogEvent
// if the devLogFunc has been set via SetDevLogFunc.
func logEvent(event *core.Event, taskType string) {
	if devLogFunc != nil {
		devLogFunc(event, taskType)
	}
}

// DevLogFunc is the signature for the dev console log callback.
type DevLogFunc func(event *core.Event, taskType string)

var devLogFunc DevLogFunc

// SetDevLogFunc sets the function used by tracked clients to log events
// to the dev console. This is called by the top-level dexcost package
// to wire up LogEvent without creating an import cycle.
func SetDevLogFunc(f DevLogFunc) {
	devLogFunc = f
}

// toResponseMap attempts to convert a response to map[string]interface{}.
func toResponseMap(resp interface{}) (map[string]interface{}, error) {
	if resp == nil {
		return nil, fmt.Errorf("nil response")
	}
	if m, ok := resp.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, fmt.Errorf("response is not map[string]interface{}")
}

File: rust/src/clients/tracked_openai.rs
//! Wrapper that records LLM cost events from OpenAI-compatible responses.
//!
//! # Usage
//!
//! ```no_run
//! use std::sync::Arc;
//! use dexcost::clients::tracked_openai::TrackedOpenAI;
//! use dexcost::core::tracker::TrackedTask;
//! use dexcost::pricing::engine::PricingEngine;
//!
//! let pricing = Arc::new(PricingEngine::new());
//! let tracked = TrackedOpenAI::new(pricing);
//!
//! // After making an OpenAI call, record the response:
//! let event = tracked.record_response("gpt-4o", 1000, 500, None, Some(250));
//! ```

use std::sync::Arc;

use crate::core::auto_task;
use crate::core::models::{CostEvent, EventType};
use crate::pricing::engine::PricingEngine;
use crate::transport::buffer::EventBuffer;

/// Wrapper that records LLM cost events from OpenAI-compatible responses.
///
/// Accepts pre-extracted token counts (no dependency on an actual OpenAI SDK
/// crate). The pricing engine is used to look up per-token costs and compute
/// the total.
pub struct TrackedOpenAI {
    pricing: Arc<PricingEngine>,
}

impl TrackedOpenAI {
    /// Creates a new `TrackedOpenAI` wrapper.
    pub fn new(pricing: Arc<PricingEngine>) -> Self {
        Self { pricing }
    }

    /// Records an OpenAI-compatible response as an LLM cost event.
    ///
    /// An auto-task is created (via [`auto_task::create_auto_task`]) for each
    /// recorded response and finalized immediately.
    ///
    /// # Arguments
    ///
    /// * `model` - The model name (e.g. `"gpt-4o"`).
    /// * `input_tokens` - Number of prompt tokens.
    /// * `output_tokens` - Number of completion tokens.
    /// * `cached_tokens` - Optional cached prompt tokens.
    /// * `latency_ms` - Optional end-to-end latency in milliseconds.
    pub fn record_response(
        &self,
        model: &str,
        input_tokens: u32,
        output_tokens: u32,
        cached_tokens: Option<u32>,
        latency_ms: Option<u64>,
    ) -> CostEvent {
        let cost_result = self.pricing.get_cost_sync(
            model,
            i64::from(input_tokens),
            i64::from(output_tokens),
            cached_tokens.map(i64::from).unwrap_or(0),
            0,
        );

        let task_type = format!("openai.{}", model);
        let mut task = crate::core::models::Task::new(&task_type);
        task.metadata
            .insert("session".to_string(), serde_json::Value::Bool(true));
        task.metadata.insert(
            "initiated_by".to_string(),
            serde_json::Value::String(task_type.clone()),
        );

        let mut event = CostEvent::new(&task.task_id, EventType::LlmCall);
        event.provider = Some("openai".to_string());
        event.model = Some(model.to_string());
        event.input_tokens = Some(i64::from(input_tokens));
        event.output_tokens = Some(i64::from(output_tokens));
        event.cached_tokens = cached_tokens.map(i64::from);
        event.latency_ms = latency_ms.map(|ms| ms as i64);
        event.cost_usd = cost_result.cost_usd;
        event.cost_confidence = cost_result.cost_confidence;
        event.pricing_source = Some(cost_result.pricing_source);
        event.pricing_version = Some(cost_result.pricing_version);

        // Log to dev console
        crate::dev_console::log_event(&event, &task.task_type);

        event
    }

    /// Records an OpenAI-compatible response and persists both the auto-task and
    /// the event into the given buffer.
    ///
    /// This is the battery-included variant that creates an auto-task, aggregates
    /// costs, and writes everything to the buffer in one call.
    pub fn record_response_buffered(
        &self,
        model: &str,
        input_tokens: u32,
        output_tokens: u32,
        cached_tokens: Option<u32>,
        latency_ms: Option<u64>,
        buffer: &mut EventBuffer,
    ) -> CostEvent {
        let event = self.record_response(
            model,
            input_tokens,
            output_tokens,
            cached_tokens,
            latency_ms,
        );

        let mut task = crate::core::models::Task::new(&format!("openai.{}", model));
        task.metadata
            .insert("session".to_string(), serde_json::Value::Bool(true));
        task.metadata.insert(
            "initiated_by".to_string(),
            serde_json::Value::String(format!("openai.{}", model)),
        );

        auto_task::finalize_auto_task(&mut task, Some(&event), "success", buffer);
        buffer.add_event(event.clone());

        event
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::models::{CostConfidence, EventType};
    use rust_decimal::Decimal;

    #[test]
    fn test_record_response_basic() {
        let pricing = Arc::new(PricingEngine::new());
        let tracked = TrackedOpenAI::new(pricing);

        let event = tracked.record_response("gpt-4o", 1000, 500, None, Some(250));

        assert_eq!(event.event_type, EventType::LlmCall);
        assert_eq!(event.provider.as_deref(), Some("openai"));
        assert_eq!(event.model.as_deref(), Some("gpt-4o"));
        assert_eq!(event.input_tokens, Some(1000));
        assert_eq!(event.output_tokens, Some(500));
        assert!(event.cached_tokens.is_none());
        assert_eq!(event.latency_ms, Some(250));
        assert!(!event.task_id.is_empty());
        assert!(!event.event_id.is_empty());
    }

    #[test]
    fn test_record_response_with_cached_tokens() {
        let pricing = Arc::new(PricingEngine::new());
        let tracked = TrackedOpenAI::new(pricing);

        let event = tracked.record_response("gpt-4o", 1000, 500, Some(200), None);

        assert_eq!(event.cached_tokens, Some(200));
        assert!(event.latency_ms.is_none());
    }

    #[test]
    fn test_record_response_unknown_model() {
        let pricing = Arc::new(PricingEngine::new());
        let tracked = TrackedOpenAI::new(pricing);

        let event = tracked.record_response("totally-unknown-model-xyz", 500, 200, None, None);

        assert_eq!(event.cost_confidence, CostConfidence::Unknown);
        assert_eq!(event.cost_usd, Decimal::ZERO);
    }

    #[test]
    fn test_record_response_buffered() {
        let pricing = Arc::new(PricingEngine::new());
        let tracked = TrackedOpenAI::new(pricing);

        let mut buffer = EventBuffer::new().unwrap();
        let event =
            tracked.record_response_buffered("gpt-4o", 1000, 500, None, Some(100), &mut buffer);

        assert_eq!(event.event_type, EventType::LlmCall);
        assert_eq!(event.provider.as_deref(), Some("openai"));
        assert_eq!(buffer.event_count(), 1);
        assert_eq!(buffer.task_count(), 1);
    }

    #[test]
    fn test_record_response_pricing_populated() {
        let pricing = Arc::new(PricingEngine::new());
        let tracked = TrackedOpenAI::new(pricing);

        let event = tracked.record_response("gpt-4o", 1000, 500, None, None);

        // gpt-4o should be a known model
        if event.cost_confidence == CostConfidence::Computed {
            assert!(event.cost_usd > Decimal::ZERO);
            assert!(event.pricing_source.is_some());
            assert!(event.pricing_version.is_some());
        }
    }
}
